### PR TITLE
SWEP: Add SWEP.InvisibleViewModel

### PR DESCRIPTION
**Pull request by EntranceJew**
- the Deploy() code is useless
- the DrawViewModel check happens each DrawHUD so it's pointless to set it once in Deploy
- the only reason it appeared to work is because the FOV of those items were set to something incredibly small

### CODE REVIEW

**Review by TimGoll — APPROVED** [review]

File: gamemodes/terrortown/entities/weapons/weapon_tttbase.lua [hunk 1]
```diff
@@ -68,6 +68,10 @@ if CLIENT then
 	-- if the files are different, it only looks at the name. I recommend you
 	-- create your own directory so that this does not happen,
 	-- eg. /materials/vgui/ttt/mycoolserver/mygun.vmt
+
+	-- The ViewModel should not draw, in any situation. Prevents the need for hacks which
+	-- overload drawing functions.
+	SWEP.InvisibleViewModel = false
```
Comment (saibotk) [inline gamemodes/terrortown/entities/weapons/weapon_tttbase.lua:74]: Which hacks are you talking about? Can you give an example?

File: gamemodes/terrortown/entities/weapons/weapon_ttt_unarmed.lua [hunk 2]
```diff
@@ -71,18 +72,6 @@ function SWEP:Reload()
...
 end
 
----
--- @ignore
-function SWEP:Deploy()
-	if SERVER and IsValid(self:GetOwner()) then
-		self:GetOwner():DrawViewModel(false)
-	end
-
-	self:DrawShadow(false)
```
Comment (saibotk) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_unarmed.lua:81]: This one had DrawShadow false too, can you please test, that all entities that did that really do not need to have this? 
Just make sure that none of them now still render a shadow / test ingame.

**Review by saibotk — CHANGES_REQUESTED** [review]

File: gamemodes/terrortown/entities/weapons/weapon_tttbase.lua (hunk 1 above)
Comment (EntranceJew) [reply gamemodes/terrortown/entities/weapons/weapon_tttbase.lua:74]: hacks as in overriding DrawViewModel in the deploy (only works once), setting the FOV < 10 (the viewmodel still exists, just out of frame)

Comment (saibotk) [reply gamemodes/terrortown/entities/weapons/weapon_tttbase.lua:74]: I still need a real example like is this even done anywhere? I would assume to see a deletion of such a hacky way in the code when you say you removed the need for it. But i cant find anything

Comment (EntranceJew) [reply gamemodes/terrortown/entities/weapons/weapon_tttbase.lua:74]: [yes, but i don't have any suggestions for what FOV an invisible viewmodel should be, so I didn't change it](https://github.com/search?q=repo%3ATTT-2%2FTTT2%20%22ViewModelFOV%20%3D%2010%22&type=code), the rest of the [DrawViewModel changes are in the changelist](https://github.com/TTT-2/TTT2/pull/1311/files)

Comment (TimGoll) [reply gamemodes/terrortown/entities/weapons/weapon_tttbase.lua:74]: > I still need a real example like is this even done anywhere?

The unarmed weapon does this, with my cam rework you are sometimes able to see the crowbar model

**Review by saibotk — APPROVED** [review]

**Comment by TimGoll** [comment]
@EntranceJew I just spent 5 hours debugging my game until I noticed that this PR breaks the view models of all weapons. If you change maps after a while no weapon has a view model anymore, only a client restart helps, not even a map change.

Any idea why? Because I don't get it

**Comment by EntranceJew** [comment]
you could change it to:
```
client:DrawViewModel(not self.InvisibleViewModel)
```

**Comment by TimGoll** [comment]
> you could change it to:
> 
> ```
> client:DrawViewModel(not self.InvisibleViewModel)
> ```

well it is already that on the master, but that isn't working

**Comment by EntranceJew** [comment]
i only had issues like this on first rounds and i attributed them to other weapon first round bugs like the slots being incorrect; [the docs](https://wiki.facepunch.com/gmod/Player:DrawViewModel) suggest that it is fine to have in a hook like that.

the alternative is to take it out of that function and move it to the base:
```lua
SWEP:PreDrawViewModel(vm, wep, ply)
  if wep.InvisibleViewModel then
    ply:DrawViewModel(false)
  end
end

SWEP:PostDrawViewModel(vm, wep, ply)
  if wep.InvisibleViewModel then
    ply:DrawViewModel(true)
  end
end
```

moving things back to the deploy/holster as we've already seen is unreliable

**Comment by TimGoll** [comment]
> i only had issues like this on first rounds

In my debug session I tested it with only TTT2 installed, no other addons. So it has to be buggy, maybe the docs are wrong here. Or this is a GMod issue.

I'll give this a try

**Comment by TimGoll** [comment]
Seems to fix the issue, I'll open a PR

**Pull request merged**